personalities/sctfe/storage.go

```diff
@@ -78,3 +87,65 @@ func (cts *CTStorage) AddIssuerChain(ctx context.Context, chain []*x509.Certific
 	}
 	return nil
 }
+
+// cachedIssuerStorage wraps an IssuerStorage, and keeps a copy the sha256 of certs it contains.
+//
+// This is intended to make querying faster. It does not keep a copy of the certs, only sha256.
+// Only up to N keys will be stored locally.
+// TODO(phboneff): add monitoring for the number of keys
+type cachedIssuerStorage struct {
+	m map[string]bool
+	N int // maximum number of entries allowed in m
+	s IssuerStorage
+}
+
+// Exists checks if the key exists in the local cache, if not checks in the underlying storage.
+// If it finds it there, caches the key locally.
+func (c cachedIssuerStorage) Exists(ctx context.Context, key []byte) (bool, error) {
+	_, ok := c.m[string(key)]
+	if ok {
+		klog.V(2).Infof("Exists: found %q in local key cache", hex.EncodeToString(key))
+		return true, nil
+	}
+	ok, err := c.s.Exists(ctx, key)
+	if err != nil {
+		return false, fmt.Errorf("error checking if issuer %q exists in the underlying IssuerStorage: %s", string(key), err)
+	}
+	if ok {
+		c.m[string(key)] = true
+	}
+	return ok, nil
+}
+
+// AddIssuers first adds the issuers to the underlying storage, then caches their sha256 locally.
+//
+// Only up to c.N issuer sha256 will be cached.
+func (c cachedIssuerStorage) AddIssuers(ctx context.Context, kv []KV) error {
+	req := []KV{}
+	for _, kv := range kv {
+		b, err := c.Exists(ctx, kv.K)
+		if err != nil {
+			return fmt.Errorf("error checking if issuer %q has been sotred previously: %v", string(kv.K), err)
+		}
+		if !b {
+			req = append(req, kv)
+		}
+	}
+	if err := c.s.AddIssuers(ctx, req); err != nil {
+		return fmt.Errorf("AddIssuers: error storing issuer data for in the underlying IssuerStorage: %v", err)
+	}
+	for _, kv := range req {
+		if len(c.m) >= c.N {
+			klog.V(2).Infof("Add: local issuer cache full, will stop caching issuers.")
+			return nil
+		}
+		c.m[string(kv.K)] = true
+	}
+	return nil
+}
+
+func NewCachedIssuerStorage(s IssuerStorage) cachedIssuerStorage {
+	c := cachedIssuerStorage{s: s, N: maxCachedIssuerKeys}
+	c.m = make(map[string]bool)
```

Comment:
nit: can inline this into the instantiation of `cachedIssuerStorage{}` above, and even inline all _that_ into the `return` :)